src/gen9_avc_encoder.c

```diff
@@ -0,0 +1,6031 @@
+    i965_free_gpe_resource(&avc_ctx->s16x_memv_data_buffer);
+    i965_free_gpe_resource(&avc_ctx->s32x_memv_data_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_brc_history_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_brc_pre_pak_statistics_output_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_brc_image_state_read_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_brc_image_state_write_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_brc_const_data_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_brc_dist_data_surface);
+    i965_free_gpe_resource(&avc_ctx->res_mbbrc_roi_surface);
+    i965_free_gpe_resource(&avc_ctx->res_mbbrc_mb_qp_data_surface);
+    i965_free_gpe_resource(&avc_ctx->res_mb_qp_data_surface);
+    i965_free_gpe_resource(&avc_ctx->res_mbbrc_const_data_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_brc_mbenc_curbe_write_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_mbenc_slice_map_surface);
+    i965_free_gpe_resource(&avc_ctx->res_sfd_output_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_sfd_cost_table_p_frame_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_sfd_cost_table_b_frame_buffer);
+    i965_free_gpe_resource(&avc_ctx->res_wp_output_pic_select_surface_list[0]);
+    i965_free_gpe_resource(&avc_ctx->res_wp_output_pic_select_surface_list[1]);
+    i965_free_gpe_resource(&avc_ctx->res_mad_data_buffer);
+
+    for(i = 0;i < 2 ; i++)
+    {
+        if (avc_ctx->wp_output_pic_select_surface_obj[i]) {
+            i965_DestroySurfaces(ctx, &avc_ctx->wp_output_pic_select_surface_id[i], 1);
+            avc_ctx->wp_output_pic_select_surface_id[i] = VA_INVALID_SURFACE;
+            avc_ctx->wp_output_pic_select_surface_obj[i] = NULL;
+        }
+    }
+
+}
+
+static void
+gen9_avc_run_kernel_media_object(VADriverContextP ctx,
+                             struct intel_encoder_context *encoder_context,
+                             struct i965_gpe_context *gpe_context,
+                             int media_function,
+                             struct gpe_media_object_parameter *param)
+{
+    struct encoder_vme_mfc_context * vme_context = (struct encoder_vme_mfc_context *)encoder_context->vme_context;
+    struct gen9_avc_encoder_context * avc_ctx = (struct gen9_avc_encoder_context * )vme_context->private_enc_ctx;
+
+    struct intel_batchbuffer *batch = encoder_context->base.batch;
+    struct encoder_status_buffer_internal *status_buffer;
+    struct gpe_mi_store_data_imm_parameter mi_store_data_imm;
+
+    if (!batch)
+        return;
+
+    intel_batchbuffer_start_atomic(batch, 0x1000);
+
+    status_buffer = &(avc_ctx->status_buffer);
+    memset(&mi_store_data_imm, 0, sizeof(mi_store_data_imm));
+    mi_store_data_imm.bo = status_buffer->bo;
+    mi_store_data_imm.offset = status_buffer->media_index_offset;
+    mi_store_data_imm.dw0 = media_function;
+    gen8_gpe_mi_store_data_imm(ctx, batch, &mi_store_data_imm);
+
+    intel_batchbuffer_emit_mi_flush(batch);
```

Comment:
Please do make sure elsewhere in the code you are flushing before you do the store, not just at this location.